lib/MySQL_HostGroups_Manager.cpp

```diff
@@ -4499,6 +4556,214 @@ void MySQL_HostGroups_Manager::read_only_action(char *hostname, int port, int re
+						host_server_mapping.set_readonly_flag(0);
+					}
+				} else {
+					// the server was already detected as RO=0
+					// no action required
+				}
+
+				if (act == true) {	// there are servers either missing, or with stats=OFFLINE_HARD
+
+					// copy all reader nodes to writer
+					host_server_mapping.copy(HostGroup_Server_Mapping::Type::WRITER, HostGroup_Server_Mapping::Type::READER);
+
+					if (mysql_thread___monitor_writer_is_also_reader) {
+						// server is also a reader, we copy all nodes from writer to reader (previous reader nodes will be reused)
+						host_server_mapping.copy(HostGroup_Server_Mapping::Type::READER, HostGroup_Server_Mapping::Type::WRITER, false);
+					} else {
+						// server can only be a writer
+						host_server_mapping.clear(HostGroup_Server_Mapping::Type::READER);
+					}
+
+					update_mysql_servers_table = true;
+				}
+			}
+		} else if (read_only == 1) {
+			if (is_writer) {
+				// the server has read_only=1 (reader), but we find it as writer, so we copy all writer nodes to reader (previous reader nodes will be reused)
+				host_server_mapping.copy(HostGroup_Server_Mapping::Type::READER, HostGroup_Server_Mapping::Type::WRITER, false);
+
+				// clearing all writer nodes
+				host_server_mapping.clear(HostGroup_Server_Mapping::Type::WRITER);
+
+				update_mysql_servers_table = true;
+			}
+		} else {
+			// LCOV_EXCL_START
+			assert(0);
+			break;
+			// LCOV_EXCL_STOP
+		}
+	}
+
+	if (update_mysql_servers_table) {
+		purge_mysql_servers_table();
+		proxy_debug(PROXY_DEBUG_MYSQL_CONNPOOL, 4, "DELETE FROM mysql_servers\n");
+		mydb->execute("DELETE FROM mysql_servers");
+		generate_mysql_servers_table();
+
+		//if (GloAdmin && GloAdmin->checksum_variables.checksum_mysql_servers) 
+		{
+			char* error = NULL;
+			int cols = 0;
+			int affected_rows = 0;
+			SQLite3_result* resultset = NULL;
+			mydb->execute_statement(MYHGM_GEN_ADMIN_RUNTIME_SERVERS, &error, &cols, &affected_rows, &resultset);
+			save_runtime_mysql_servers(resultset); // assigning runtime_mysql_servers with updated mysql server resultset 
+			
+			resultset = NULL;
+
+			char* query = (char*)"SELECT hostgroup_id, hostname, port, gtid_port, CASE status WHEN 0 OR 1 OR 4 THEN 0 ELSE status END status, weight, compression, max_connections, max_replication_lag, use_ssl, max_latency_ms, comment FROM mysql_servers WHERE status<>3 ORDER BY hostgroup_id, hostname, port";
+			mydb->execute_statement(query, &error, &cols, &affected_rows, &resultset);
+
+			if (resultset) {
+				if (resultset->rows_count) {
+					table_resultset_checksum[HGM_TABLES::MYSQL_SERVERS] = resultset->raw_checksum();
+
+					hgsm_mysql_servers_checksum = table_resultset_checksum[HGM_TABLES::MYSQL_SERVERS];
+				}
+				delete resultset;
+			} else {
+				proxy_info("Checksum for table %s is 0x%lX\n", "mysql_servers", (long unsigned int)0);
+			}
+
+			uint64_t hash = 0, hash2 = 0;
+			SpookyHash myhash;
```

Comment:
This checksum computation logic is different from the one found in `commit`. This can lead to different checksums when generated via `read_only_action_v2` and via `commit`, and prevent successful synchronizations of cluster nodes. In `commit` the classical logic is only initializing the `SpookyHash` if the `resultset` holds any value, here it's initialized and used no matter if the module checksum is `0` (meaning that the `resultset` from which it was computed has no results). Since the `table_resultset_checksum` is already conveniently arranged in the proper order for the checksum computation, for mimicking the old logic from commit should be enough to do something like the following:

```
			SpookyHash myhash;
			bool init = false;

			for (uint64_t hash_val : table_resultset_checksum) {
				if (hash_val) {
					if (init == false) {
						init = true;
						myhash.Init(19, 3);
					}

					myhash.Update(&hash_val, sizeof(hash_val));
				}
			}

			myhash.Final(&hash, &hash2);
```

Also, the `table_resultset_checksum` should be zero-initalized, replacing the current definition with:

```
	std::array<uint64_t, __HGM_TABLES_SIZE> table_resultset_checksum { 0 };
```

Reply:
Done